### PR TITLE
Use separate classes for error formatters

### PR DESCRIPTION
**Pull request by shadowhand**
As per middlewares/json-exception-handler#1, modifies this package to be a general purpose error catching middleware with various formatting options.

This proposes *significant* changes to the middleware and will be a new major version.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,95 +7,66 @@
...
-
-        //Any output is captured and added to the response's body
-        echo $error->getMessage();
-
-        return (new Response())->withStatus($error->getCode());
-    }
-}
-
+use Middlewares\ErrorHandler;
+use Middlewares\Formatter;
```
Comment (oscarotero) [inline README.md]: I think `Middlewares\Formatter` should be moved to `Middlewares\ErrorHandler\Formatter`, to avoid possible conflicts with other middlewares' packages

File: src/Formatter/FormatterInterface.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+namespace Middlewares\Formatter;
+
+use Throwable;
+
+interface FormatterInterface
+{
+    /**
+     * Get the output content type
+     */
+    public function contentType(): string;
```
Comment (oscarotero) [inline src/Formatter/FormatterInterface.php]: this should return an array of strings, in order to be used for various content types. For example, json error handler could be used by `application/json`, `application/manifest+json`, `application/vnd.geo+json` and so on.

File: composer.json [hunk 3]
```diff
@@ -8,19 +8,24 @@
...
+        "formatter",
+        "handler"
     ],
     "homepage": "https://github.com/middlewares/error-handler",
     "support": {
         "issues": "https://github.com/middlewares/error-handler/issues"
     },
     "require": {
-        "php": "^7.0",
+        "php": "^7.1",
```
Comment (oscarotero) [inline composer.json]: Is there any reason to increment the minimum php version supported? All middlewares support 7.0 (like PSR-15 interfaces), and I don't want to change it without a strong reason.

File: src/ErrorHandler.php [hunk 4]
```diff
@@ -1,120 +1,93 @@
...
+    private $formatters = [];
 
-        return $this;
+    public function __construct(
+        ResponseFactoryInterface $responseFactory = null,
+        StreamFactoryInterface $streamFactory = null
+    ) {
+        $this->responseFactory = $responseFactory ?? Factory::getResponseFactory();
+        $this->streamFactory = $streamFactory ?? Factory::getStreamFactory();
     }
```
Comment (oscarotero) [inline src/ErrorHandler.php:34]: I prefer to use methods to configure the factories instead inject them in the constructor, for api consistency with the rest of middlewares. There are Traits in middlewares/utils for this purpose: https://github.com/middlewares/utils/tree/master/src/Traits.

File: src/ErrorHandler.php [hunk 5]
```diff
@@ -1,120 +1,93 @@
...
-     */
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        ob_start();
-        $level = ob_get_level();
-
         try {
-            $response = $handler->handle($request);
-
-            if ($this->isError($response->getStatusCode())) {
```
Comment (oscarotero) [inline src/ErrorHandler.php:77]: Why did you remove the ability to handle responses with error? (status code between 400-599)
The purpose of this middleware is to handle all http errors, not only 500.

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (shadowhand) [reply README.md]: Makes sense. 👍

File: src/Formatter/FormatterInterface.php (hunk 2 above)
Comment (shadowhand) [reply src/Formatter/FormatterInterface.php]: This is for the output content type. I can add a separate `acceptTypes(): array` method to handle input. **However** accept type is often more complicated than just a simple list of matches. Maybe it should be a regex instead? Eg:

```php
public function acceptTypes(): string
{
    return 'application/*json*';
}
```

File: src/ErrorHandler.php (hunk 4 above)
Comment (shadowhand) [reply src/ErrorHandler.php:34]: I strongly dislike that approach because it means that `$this->responseFactory` is always in an unknown state. There is no assurance that it is actually a `ResponseFactoryInterface`, because it won't be created until either `responseFactory()` or `createResponse()` is called. 😕 

Also, the purpose of a constructor is to define things that the object needs to complete work. In this case, it *needs* a stream and response factory.

File: src/ErrorHandler.php (hunk 5 above)
Comment (shadowhand) [reply src/ErrorHandler.php:77]: And what is the definition of "http error"? What value does blowing away an existing response with an error status provide? For instance, if my API already returns 404 based on routing, why would I want to lose that and have it replaced with some generic error message?

Comment (shadowhand) [reply src/ErrorHandler.php:77]: Not to mention things like [HTTP Problem](https://docs.zendframework.com/zend-problem-details/response/) that can be used to define a PSR-7 response with a specific error format.

File: composer.json (hunk 3 above)
Comment (shadowhand) [reply composer.json]: [safe](https://packagist.org/packages/thecodingmachine/safe) requires PHP >= 7.1.

PHP 7.0 will be unsupported on 3 Dec 2018, which is nearly here. Not a lot of value in supporting it, IMO.

File: src/Formatter/FormatterInterface.php (hunk 2 above)
Comment (oscarotero) [reply src/Formatter/FormatterInterface.php]: middlewares/negotiator reduces the `Accept` header values to a single and most accurate value, that can be used in the `Content-Type`. So a request with `application/json` returns a `application/json` response, and `application/manifest+json` returns a `application/manifest+json`.

The same error handler could return responses with different content type, based in the `Accept` header. If we delegate the heavy work to middlewares/negotiator, the code to include here is trival.

Regex is an option, but may be a bit complicated for simple use cases.

File: composer.json (hunk 3 above)
Comment (oscarotero) [reply composer.json]: I don't think it's a good idea to include a dependency only [to execute this code](https://github.com/thecodingmachine/safe/blob/78a5636c9dbf03630c15036f1ee2cb78eeb09239/generated/json.php#L40) and that force us to increment the minimum php version required. Besides, there may be users that install this package and they do not even use that formatter.

File: src/ErrorHandler.php (hunk 4 above)
Comment (oscarotero) [reply src/ErrorHandler.php:34]: This topic was discused here: https://github.com/middlewares/fast-route/pull/9#issuecomment-409985531. The factories are not always needed, they are optional. There are many middlewares requiring factories, some for trivial purposes (for example, create redirect responses only [under specific conditions](https://github.com/middlewares/trailing-slash/blob/master/src/TrailingSlash.php#L53) and others that make a more expensive use of factories (like [read and serve static files](https://github.com/middlewares/filesystem/blob/master/src/Reader.php)).
I try to maintain all these middlewares from more global point of view, in which they share some kind of consistency in the API.

File: src/ErrorHandler.php (hunk 5 above)
Comment (oscarotero) [reply src/ErrorHandler.php:77]: I have some webpages using this feature to display generic 404, 500, etc error pages.
A workaround could be an option to enable this feature (disabled by default), allowing to define the status code that will be considered error.

File: src/ErrorHandler.php [hunk 6]
```diff
@@ -1,120 +1,93 @@
...
+    protected function errorResponse(FormatterInterface $formatter, Throwable $e): ResponseInterface
     {
-        $request = $request->withAttribute($this->attribute, $exception);
-        $handler = $this->handler ?: new ErrorHandlerDefault();
+        $responseBody = $this->streamFactory->createStream($formatter->format($e));
 
-        return $handler->handle($request);
+        $response = $this->responseFactory->createResponse($this->errorStatus($e));
+        $response = $response->withHeader('Content-Type', $formatter->contentType());
+        $response = $response->withBody($responseBody);
```
Comment (oscarotero) [inline src/ErrorHandler.php:82]: Why create a stream outside of the response instead simply doing a `$response->getBody()->write($formatter->format($e))`? That would remove the dependency of streamFactory.

File: src/ErrorHandler.php (hunk 4 above)
Comment (shadowhand) [reply src/ErrorHandler.php:34]: Easier doesn't make it right. Not knowing the dependencies makes it harder to use, IMO. Especially if you use auto-wiring as a dependency injection strategy.

For me this is kind of a deal breaker. I really don't want to dilute my code and make it harder to use with auto-wiring. I totally understand where you are coming from with the trait, but all it does is shift the burden of managing dependencies from you to everyone else. That's a 👎 in my book.

I would ask you to reconsider this stance or I am likely going to put my efforts in a different direction.

File: src/ErrorHandler.php (hunk 5 above)
Comment (shadowhand) [reply src/ErrorHandler.php:77]: That doesn't seem like the right approach to me.

File: src/ErrorHandler.php (hunk 4 above)
Comment (oscarotero) [reply src/ErrorHandler.php:34]: Ok, maybe you're right. I don't use auto-wiring and didn't consider the factories so important that must to be passed through the constructor (actually they can be configured in the static Factory class or using the methods). But you're the second person claiming for auto wiring, so I guess I was wrong.

File: src/ErrorHandler.php (hunk 5 above)
Comment (oscarotero) [reply src/ErrorHandler.php:77]: Http errors can come from many places:
- The router
- The controller
- A middleware performing some kind of authentication/authorization
- A middleware with a firewall
- A body parser trying to parse a malformed body
- An antispam middleware
- Cors
- etc...

The initial purpose of this middleware was handling all these errors in an unique place to, for example, generate the same 404 page error independently of where the error was originated. I don't want to lose this functionality (that I already have used in some projects). If you want to limit this middleware to only catch exceptions and convert them to 500 responses, then I need to create a http-error-handler middleware.

**Review by oscarotero — APPROVED** [review]

File: src/ErrorHandler.php (hunk 6 above)
Comment (shadowhand) [reply src/ErrorHandler.php:82]: `getBody()->write()` is not idempotent for any number of reasons:

- body is read-only
- body already has content of a different type
- possibly others

It is far safer to always replace the body than attempt to write into it.

File: src/ErrorHandler.php (hunk 5 above)
Comment (shadowhand) [reply src/ErrorHandler.php:77]: Again, I don't see the point of that. All of those cases (assuming they are not exceptional) should already be returning a fully-formed response with details about the error. Each of the examples listed should tell the consumer *why* the request failed. And then I go back to: why would we want to overwrite a fully-formed error response with a generic response?

File: src/ErrorHandler.php (hunk 6 above)
Comment (oscarotero) [reply src/ErrorHandler.php:82]: If the response has just been created (so I assume the body too), can the body already contain content or be read-only? That wouldn't be compliance with psr-17.

File: src/ErrorHandler.php (hunk 5 above)
Comment (oscarotero) [reply src/ErrorHandler.php:77]: There may be cases where many different responses display the same result. An example is github: you see the same 404 error page if you enter in an inexistent user/repository (https://github.com/asofijasfo) or if the repository exists but you don't have permission to see, or are not logged. 

Sometimes I need to show the same 404 page over and over, but all there responses are created from many sites:

- The route has not found
- The route has found but the object was not found in the database
- The object has found but it's not visible or the user has not permission
- Etc.

In addition to that, there's the [HttpErrorException](https://github.com/middlewares/error-handler/blob/master/src/HttpErrorException.php) created to convert any exception into http errors. That allows to access to the exception that causes the http error in the error handler and do something with it (logging, display info to the user, etc).

Comment (shadowhand) [reply src/ErrorHandler.php:77]: Right, I would expect HttpErrorException to be thrown in any situation where you want a generic 404 page.

File: composer.json (hunk 3 above)
Comment (shadowhand) [reply composer.json]: I disagree, but not worth fighting about.

File: src/ErrorHandler.php (hunk 6 above)
Comment (shadowhand) [reply src/ErrorHandler.php:82]: In theory, yes. That's not enforceable though, so I would rather write code that is assuredly safe.

**Comment by shadowhand** [comment]
@oscarotero all requested changes made.
